File: scala/src/main/scala/math/Algebra.scala
package rays.math

import scala.math.sqrt

/** Returns the solution of a linear equation.
  *
  * A linear equation is a first-prder polinomial equation of the form `ax + b = 0`.
  */
def solveLinear(a: Double, b: Double): Option[Double] =
  Option.unless(a == 0)(-b / a)

/** Returns the roots of a quadratric equation.
  *
  * A quadratic equation is a second-order polynomial equation of the form `ax^2 + bx + c = 0`.
  */
def solveQuatratic(a: Double, b: Double, c: Double): Option[(Double, Double)] =
  val discriminant = (b * b) - (4 * a * c)
  Option.when(discriminant >= 0d):
    if discriminant == 0.0 then
      // Only one real root.
      val x = -0.5 * b / a
      (x, x)
    else
      val q = -0.5 * (b + (if b > 0d then 1 else -1) * sqrt(discriminant))
      val (x0, x1) = (q / a, c / q)
      if x0 < x1 then (x0, x1) else (x1, x0)

File: swift/Rays/Inheritance/Math/Algebra.swift
import RealModule

/// Returns the solution of a linear equation.
///
/// A linear equation is a first-prder polinomial equation of the form `ax + b = 0`.
public func solveLinear(a: Double, b: Double) -> Double? {
  guard a != 0
    else { return nil }
  return -b / a
}

/// Returns the roots of a quadratric equation.
///
/// A quadratic equation is a second-order polynomial equation of the form `ax^2 + bx + c = 0`.
public func solveQuatratic(a: Double, b: Double, c: Double) -> (Double, Double)? {
  let discriminant = (b * b) - (4 * a * c)

  if discriminant == 0.0 {
    // Only one real root.
    let x = -0.5 * b / a
    return (x, x)
  } else if discriminant > 0.0 {
    let q = b > 0.0
      ? -0.5 * (b + Double.sqrt(discriminant))
      : -0.5 * (b - Double.sqrt(discriminant))

    let x0 = q / a
    let x1 = c / q
    return x0 < x1
      ? (x0, x1)
      : (x1, x0)
  } else {
    return nil
  }
}
